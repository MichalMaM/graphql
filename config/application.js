// @flow

import url from 'url';

export default {
  restApiEndpoint: {
    allBookings:
      'https://booking-api.skypicker.com/api/v0.1/users/self/bookings',
    singleBooking: (bookingId: number, simpleToken: string) =>
      queryWithParameters(
        `https://booking-api.skypicker.com/api/v0.1/users/self/bookings/${bookingId}`,
        {
          simple_token: simpleToken,
        },
      ),
    allFlights: (queryParameters: ?Object = null) =>
      queryWithParameters('https://api.skypicker.com/flights', {
        ...queryParameters,
        v: 3,
      }),
    login: 'https://auth.skypicker.com/v1/user.login',
    identity: 'https://auth.skypicker.com/v1/user.get',
    airlines: 'https://api.skypicker.com/airlines?v=2',
    allLocations: (queryParameters: Object) =>
      queryWithParameters(
        'https://api.skypicker.com/locations',
        queryParameters,
      ),
    rates: 'https://api.skypicker.com/rates',
    bags: (bookingId: number) =>
      queryWithParameters(
        `https://booking-api.skypicker.com/mmb/v1/bookings/${bookingId}/bags`,
      ),
  },
  auth: {
    basicToken: String(process.env.AUTH_BASIC_TOKEN),
    digest: String(process.env.AUTH_DIGEST),
  },
};

export function queryWithParameters(
  absoluteUrl: string,
  queryParameters: ?Object = null,
) {
  const urlObject = url.parse(absoluteUrl, true);
  // intentional ==, can be null or undefined
  if (queryParameters != null) {
    urlObject.query = filterParameters(queryParameters);
  }
  return url.format(urlObject);
}

function filterParameters(parameters: Object) {
  const filteredParameters = parameters;
  Object.keys(parameters).forEach(key => {
    const val = parameters[key];
    if (val === null || val === undefined || val === '') {
      delete filteredParameters[key];
      return;
    }

    if (Array.isArray(val)) {
      filteredParameters[key] = val.toString();
    }

    if (typeof val === 'string') {
      filteredParameters[key] = val.trim();
    }
  });
  return filteredParameters;
}
