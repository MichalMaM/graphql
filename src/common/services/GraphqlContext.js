// @flow

import DataLoader from 'dataloader';
import idx from 'idx';
import sha1 from 'sha1';

import parseAcceptLanguage from './context/ParseAcceptLanguage';

import IdentityDataloader from '../../identity/dataloaders/Identity';
import createBookingLoader from '../../booking/dataloaders/Booking';
import createCurrencyLoader from '../../currency/dataloaders/Currency';
import createAirlineLoader from '../../flight/dataloaders/Airline';
import createRatesLoader from '../dataloaders/Rates';
import createFAQLoader from '../../FAQ/dataloaders/searchFAQ';
import FAQCategoriesLoader from '../../FAQ/dataloaders/FAQCategories';
import createFAQArticleLoader from '../../FAQ/dataloaders/FAQArticle';
import createSingleBookingLoader from '../../booking/dataloaders/SingleBooking';
import BookingsLoader from '../../booking/dataloaders/Bookings';
import LocationsLoader from '../../location/dataloaders/Locations';
import LocationLoader from '../../location/dataloaders/Location';
import FlightLoader from '../../flight/dataloaders/Flight';
import OptionsStorage from './context/OptionsStorage';
import HotelsAvailability from '../../hotel/dataloaders/HotelsAvailability';
import { type SearchParameters as HotelKey } from '../../hotel/dataloaders/flow/SearchParameters';
import createHotelByIdLoader from '../../hotel/dataloaders/HotelByID';
import HotelCities from '../../hotel/dataloaders/HotelCities';
import CitiesByID from '../../hotel/dataloaders/CitiesByID';
import HotelRoomsLoader from '../../hotel/dataloaders/HotelRooms';
import HotelRoomAvailabilityLoader from '../../hotel/dataloaders/HotelRoomAvailability';
import HotelRoomBeddingLoader from '../../hotel/dataloaders/RoomBedding';
import PriceStatsLoader from '../../hotel/dataloaders/PriceStats';
import DynamicPackagesLoader from '../../dynamicPackage/dataloaders/DynamicPackages';
import createCustomerSupportNumberLoader from '../../customerSupport/dataloaders/CustomerSupportNumber';

import type { Booking } from '../../booking/Booking';
import type { Args as SingleBookingArgs } from '../../booking/dataloaders/SingleBooking';
import type { Airline } from '../../flight/Flight';
import type { CurrencyDetail } from '../../currency/CurrencyDetail';
import type { HotelType } from '../../hotel/dataloaders/flow/HotelType';
import type { City } from '../../hotel/dataloaders/flow/City';
import type { Args as FAQArgs } from '../../FAQ/dataloaders/searchFAQ';
import type { FAQArticleItem } from '../../FAQ/dataloaders/FAQCategories';
import type {
  Args as FAQCArticleArgs,
  FAQArticleDetail,
} from '../../FAQ/dataloaders/FAQArticle';
import type { CustomerSupportNumber } from '../../customerSupport/types/CustomerSupportNumber';

/**
 * FIXME:
 * This is stupid - it's already defined by data-loader itself but currently
 * it's needed to type resolvers. How to do it better?
 */
export type GraphqlContextType = {|
  // DataLoader<K, V>
  locale: {|
    +language: string, // cs
    +territory: string, // CZ
    +format: {|
      +underscored: string, // cs_CZ
      +dashed: string, // cs-CZ
    |},
  |},
  apiToken: ?string,
  whoIAm: string,
  dataLoader: {|
    airline: DataLoader<string, ?Airline>,
    booking: DataLoader<number | string, Booking>,
    bookings: BookingsLoader,
    singleBooking: DataLoader<SingleBookingArgs, Booking>,
    currency: DataLoader<string, CurrencyDetail>,
    flight: FlightLoader,
    identity: IdentityDataloader,
    location: LocationLoader,
    locations: LocationsLoader,
    rates: DataLoader<string, ?number>,
    city: DataLoader<string, City>,
    hotel: {
      availabilityByLocation: DataLoader<HotelKey, HotelType[]>,
      availabilityByID: DataLoader<HotelKey, HotelType[]>,
      byID: DataLoader<$FlowFixMe, $FlowFixMe>,
      cities: HotelCities,
      room: HotelRoomsLoader,
      roomAvailability: HotelRoomAvailabilityLoader,
      roomBedding: typeof HotelRoomBeddingLoader,
      priceStats: DataLoader<
        { searchParams: HotelKey, boundary: 'MAX' | 'MIN' },
        number,
      >,
    },
    FAQ: DataLoader<FAQArgs, FAQArticleItem[]>,
    FAQCategories: FAQCategoriesLoader,
    FAQArticle: DataLoader<FAQCArticleArgs, FAQArticleDetail>,
    dynamicPackages: DynamicPackagesLoader,
    customerSupportNumber: DataLoader<string, ?CustomerSupportNumber>,
  |},
  options: OptionsStorage,
  _traceCollector?: Object,
|};

/**
 * Although API contains several FAQ trees, this is currently the only relevant tree.
 * Others may include testing data or the ones not intended for displaying
 */
const FAQ_CATEGORY_ID = '3';

type ContextParameterType = {
  token?: ?string,
  acceptLanguage?: ?string,
  request: {
    ip: string,
    headers: Object,
  },
};

export function createContext({
  token,
  acceptLanguage,
  request,
}: ContextParameterType): GraphqlContextType {
  const bookings = new BookingsLoader(token);
  const locations = new LocationsLoader();
  const location = new LocationLoader();
  const hotelCities = new HotelCities();

  const [language, territory] = parseAcceptLanguage(acceptLanguage);
  const userAgent = idx(request, _ => _.headers['user-agent']) || '';
  const userIp = idx(request, _ => _.ip) || '';

  const locale = {
    language,
    territory: territory,
    format: {
      underscored: language + '_' + territory,
      dashed: language + '-' + territory,
    },
  };

  return {
    locale,
    apiToken: token,
    whoIAm: sha1(userAgent + userIp),
    dataLoader: {
      airline: createAirlineLoader(),
      booking: createBookingLoader(token, bookings),
      bookings: bookings,
      singleBooking: createSingleBookingLoader(),
      currency: createCurrencyLoader(),
      flight: new FlightLoader(location, locations),
      identity: new IdentityDataloader(token),
      location: location,
      locations: locations,
      rates: createRatesLoader(),
      city: CitiesByID,
      hotel: {
        availabilityByLocation: HotelsAvailability,
        availabilityByID: HotelsAvailability,
        byID: createHotelByIdLoader(locale.format.underscored),
        cities: hotelCities,
        room: new HotelRoomsLoader(locale.format.underscored),
        roomAvailability: new HotelRoomAvailabilityLoader(),
        roomBedding: HotelRoomBeddingLoader,
        priceStats: PriceStatsLoader,
      },
      FAQ: createFAQLoader(locale.language, FAQ_CATEGORY_ID),
      FAQCategories: new FAQCategoriesLoader(locale.language, FAQ_CATEGORY_ID),
      FAQArticle: createFAQArticleLoader(locale.language),
      dynamicPackages: new DynamicPackagesLoader(location),
      customerSupportNumber: createCustomerSupportNumberLoader(),
    },
    options: new OptionsStorage(),
  };
}
