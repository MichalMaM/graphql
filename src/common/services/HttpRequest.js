// @flow

import { URL } from 'url';
import { fetchJson } from './JsonFetcher';

export async function get(
  absoluteApiUrl: string,
  token: ?string,
  requestHeaders?: Object = {},
  brand: ?string,
): Promise<Object> {
  const urlObject = new URL(absoluteApiUrl);

  if (token !== null && token !== undefined) {
    urlObject.searchParams.append('token', token);
  }
  if (brand !== null && brand !== undefined) {
    urlObject.searchParams.append('brand', brand);
  }
  const headers = {
    'X-Client': 'graphql',
    'Content-Type': 'application/json',
    ...requestHeaders,
  };

  return fetchJson(urlObject.toString(), 'GET', { headers });
}

export async function post(
  absoluteApiUrl: string,
  payload: Object,
  requestHeaders?: Object = {},
): Promise<Object> {
  const body = JSON.stringify(payload);
  const headers = {
    'Content-Type': 'application/json',
    'X-Client': 'graphql',
    ...requestHeaders,
  };

  return fetchJson(absoluteApiUrl, 'POST', { body, headers });
}
