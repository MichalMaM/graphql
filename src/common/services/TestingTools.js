// @flow

import {
  graphql as originalGraphQL,
  validate as originalValidate,
  parse,
} from 'graphql';
import schema from '../../Schema';
import { createContext } from './GraphqlContext';

/**
 * TestingTools must be imported FIRST otherwise mock of HttpRequest won't work.
 */
jest.mock('./HttpRequest');
jest.mock('../../hotel/services/BookingComRequest');
const mockedHttpRequest = require('./HttpRequest');
const mockedBookingComRequest = require('../../hotel/services/BookingComRequest');

class MockResponse {
  absoluteUrl: string;
  httpMethod: string;

  constructor(httpMethod: 'GET' | 'POST' | 'PATCH', absoluteUrl: string) {
    this.httpMethod = httpMethod;
    this.absoluteUrl = absoluteUrl;
  }

  replyWithData(response: Array<Object> | Object) {
    // $FlowAllowMockMagic
    mockedHttpRequest.__setMockData(
      this.httpMethod,
      this.absoluteUrl,
      response,
    );
  }
}

class MockBookingComResponse extends MockResponse {
  replyWithData(response: Array<Object> | Object) {
    // $FlowAllowMockMagic
    mockedBookingComRequest.__setMockData(
      this.httpMethod,
      this.absoluteUrl,
      response,
    );
  }
}

export class RestApiMock {
  static onGet(absoluteUrl: string) {
    return new MockResponse('GET', absoluteUrl);
  }

  static onPost(absoluteUrl: string) {
    return new MockResponse('POST', absoluteUrl);
  }
  static onPatch(absoluteUrl: string) {
    return new MockResponse('PATCH', absoluteUrl);
  }
}

export class BookingComApiMock {
  static onGet(absoluteUrl: string) {
    return new MockBookingComResponse('GET', absoluteUrl);
  }
}

export const graphql = async (
  query: string,
  variables: ?Object,
  acceptLanguage: ?string,
): Promise<Object> =>
  originalGraphQL(
    schema,
    query,
    null,
    createContext({
      token: 'test_token',
      acceptLanguage,
      request: {
        ip: '',
        headers: {},
      },
    }),
    variables,
  );

export const validate = (query: string) => {
  return originalValidate(
    schema,
    parse(query, {
      noLocation: true,
    }),
  );
};

export const evaluateResolver = (
  field: mixed,
  testValue: mixed,
  argsValue?: Object,
  contextValue?: Object,
) => {
  // $FlowAllowNextLineInThisTest (fields are possibly undefined but we assume it's fine)
  return field.resolve(testValue, argsValue, contextValue);
};
