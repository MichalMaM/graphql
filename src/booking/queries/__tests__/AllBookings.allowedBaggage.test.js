// @flow

import { graphql, RestApiMock } from '../../../common/services/TestingTools';
import config from '../../../../config/application';
import AllBookingsDataset from '../../datasets/AllBookings.json';
import Booking2707224Dataset from '../../datasets/booking-2707224.json';
import Booking2707229Dataset from '../../datasets/booking-2707229.json';
import Booking2707251Dataset from '../../datasets/booking-2707251.json';

const { allBookings } = config.restApiEndpoint;

beforeEach(() => {
  RestApiMock.onGet(allBookings).replyWithData(AllBookingsDataset);
  RestApiMock.onGet(
    `${allBookings}/2707251\\?simple_token=[0-9a-f-]{36}`,
  ).replyWithData(Booking2707251Dataset);
  RestApiMock.onGet(
    `${allBookings}/2707229\\?simple_token=[0-9a-f-]{36}`,
  ).replyWithData(Booking2707229Dataset);
  RestApiMock.onGet(
    `${allBookings}/2707224\\?simple_token=[0-9a-f-]{36}`,
  ).replyWithData(Booking2707224Dataset);
});

it('should return baggage parameters', async () => {
  const baggageQuery = `{
      allBookings {
        edges {
          node {
            databaseId
            allowedBaggage {
              cabin { height, length, width, weight, note }
              checked { height, length, width, weight, note }
            }
          }
        }
      }
    }`;
  expect(await graphql(baggageQuery)).toMatchSnapshot();
});
