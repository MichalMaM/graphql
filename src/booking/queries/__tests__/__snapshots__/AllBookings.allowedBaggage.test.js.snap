// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`should return baggage parameters 1`] = `
Object {
  "data": Object {
    "allBookings": Object {
      "edges": Array [
        Object {
          "node": Object {
            "allowedBaggage": Object {
              "cabin": Array [
                Object {
                  "height": 25,
                  "length": 55,
                  "note": null,
                  "weight": 8,
                  "width": 45,
                },
              ],
              "checked": Array [
                Object {
                  "height": 40,
                  "length": 100,
                  "note": null,
                  "weight": 23,
                  "width": 59,
                },
              ],
            },
            "databaseId": 2707224,
          },
        },
        Object {
          "node": Object {
            "allowedBaggage": Object {
              "cabin": Array [
                Object {
                  "height": 20,
                  "length": 55,
                  "note": null,
                  "weight": 7,
                  "width": 40,
                },
              ],
              "checked": Array [
                Object {
                  "height": 81,
                  "length": 119,
                  "note": null,
                  "weight": 15,
                  "width": 119,
                },
              ],
            },
            "databaseId": 2707229,
          },
        },
        Object {
          "node": Object {
            "allowedBaggage": Object {
              "cabin": Array [
                Object {
                  "height": 20,
                  "length": 55,
                  "note": null,
                  "weight": 10,
                  "width": 40,
                },
              ],
              "checked": Array [
                Object {
                  "height": 79,
                  "length": 250,
                  "note": "max 180cm (length + width + height)",
                  "weight": 23,
                  "width": 112,
                },
              ],
            },
            "databaseId": 2707251,
          },
        },
      ],
    },
  },
}
`;
